### DISCUSSION TITLE
Shutdown issue on Linkstation LS-GL series on Buster/Bullseye

### DISCUSSION THREAD

**Discussion by DanOnDeb** [1000001101000/Debian_on_Buffalo#99] [Q&A]
Hi everybody, 

I have spent quite some time on this issue so far, but cannot seem to get anywhere. 

I call two Buffalo Linkstations my own. Both older ones, and both LS-GL series.

I have been running Debian on them for quite a long time now and was always really happy with it (started on Lenny, I believe), even though the hardware is a bit outdated. As I use them for automated backup, I don't need much performance. 

I had Debian Stretch running without any issues and also upgrading to Buster was not an issue at all. However, ever since the update to Buster and the new kernel, the machine doesn't shutdown properly anymore. 

Instead, on calling "halt", "shutdown", or "poweroff", the Linkstation seems to be rebooting into emergency mode. Constant "neenaa neenaa" and the error LED is flashing. The Linkstation is not ping-able in this mode and I am assuming it has rebooted into TFTP mode at this stage (cannot verify). After a couple of minutes, the machine shuts down but is still constantly beeping until I push the power-button.

Here's what I have tried so far:
1. Upgraded to Bullyeye to see if the issue is Buster-related --> unfortunately not, same issue on Debian 11
2. Upgraded my other Linkstation to see if it was the LS is causing the issue --> unfortunately, now both have the same issue (one on Buster, one on Bullseye)
3. Checked all log files known to me, but there is nothing strange that pops up when I do. The reboot doesn't show up in the Debian logs (as it wouldn't if the LS is actually in emergency mode)
4. Searched the web and the forum here and discovered the WoL-related shutdown issue on newer LS, but mine does not have the WoL features, AFAIK. https://github.com/1000001101000/Debian_on_Buffalo/wiki/Device-Specific-Notes
5. Almost gave up on the matter. But decided to ask here.

Any feedback is very much appreciated. I am happy to supply other input and am also happy to test anything on my device. As I said, upgrading to Bullseye was no problem at all.

Thank you very much for keeping old hardware alive.
I really appreciate the effort.

Thanks, DanOnDeb

**Comment by 1000001101000** [comment]
Good to hear from you!

I dusted off my LS-GL and upgraded the Buster install that was on there to the latest version including the 4.19.0-16-marvell kernel. Mine handled shutdown and restart properly without going into EM mode or setting off any alarms.

I think I have a pretty good idea what is going on with your device and have some ideas on how you can correct it.

These devices have an on-board microcontroller which handles a bunch of features such as power management. I've created some crude documentation of the protocol used to communicate with it here:
https://buffalonas.miraheze.org/wiki/Terastation_Microcontroller_Interface

I also put together a crude python library to send signals to it (before I released there was already a tool for that):
https://github.com/1000001101000/Python_buffalo_libmicon

To shutdown this device successfully you must send it a signal to indicate you want it to poweroff. The script I use to accomplish this can be found here:
https://github.com/1000001101000/Debian_on_Buffalo/blob/master/Tools/micon_scripts/micon_shutdown.py
If you use my installer this script is configured to run as part of the shutdown process but you could also set it up manually.

Previously these signals were handled by a program called micro-evtd which will have been setup with your original Debian install which normally starts at startup and runs some sort of daemon to handle those signals as well as the fan speed and a few other things. I'm guessing that it isn't working properly anymore (I'm not sure if it's being maintained or not).

.... I started getting off topic talking for a minute there... It was an interesting project to work on ...

In short, to get this working properly again you can:
1. Install python3 and python3-serial
2. download my shutdown script and copy it to /lib/systemd/system-shutdown/ and mark it as executable.
(I'm assuming you're using systemd).

Another option would be to do a fresh install using my installer which would set that up fro you automatically.

Let me know if you have any questions!

**Reply by DanOnDeb** [reply]
Hi 1000001101000,

First off, thank you so much for the quick response. I really appreciate the effort and your dedication here.

Unfortunately, your solution did not fix it for me, but I think I may have found the reason why. 

I think I'm still running init instead of systemd. The system is not a fresh install but instead I have upgraded multiple times from my initial conversion from stock to Lenny. This may be the reason for the system still using init. As it always worked so far (by the way, the fan control works just fine), I have never bothered to do a clean fresh install.

```
root@tatooine:~# ps --no-headers -o comm 1
init

```
If you have an idea how to fix the issue using init or know an easy way to convert the running system to systemd without bricking the device, I would really like to know and try it out. If you don't, I will gladly try your installer script as soon as I have a couple of hours at hand. 

Thanks again for your help!

73,
DanOnDeb

**Reply by 1000001101000** [reply]
The script needs to run very late in the shutdown process since the microcontroller will perform that actual poweroff. I spent a little time reading documentation about sysvinit shutdown but I didn't see a good anser on how to do that. Based on what I read on my distant memories of working on other systems it would probably needed to be called as the last script for the appropriate run level (0?). Sadly, I don't have any init-based systems handy to try it out on.

I imagine with some reading and trial/error you can figure something out. You'd also need to either modify the script or pass it "halt" as a parameter since systemd passes it the mode as an argument (restart/halt/kexec/etc).

If it was me I would do a fresh install, but that is always my preference. You should be able to do it without losing your data partition though, as always, make sure you have a backup before trying it.

**Reply by DanOnDeb** [reply]
Thanks again for the and compelling feedback.

Here's a quick update:
Turns out, switching to systemd is much easier than I had thought. Since I can tinker with the one LS, I just followed the instructions on https://wiki.debian.org/systemd and skipped the part where I insert systemd as an option into the boot loader, and did:

```
apt install systemd
apt install systemd-sysv
```
and after a reboot, `ps` reveals:
```
root@tatooine:~# ps --no-headers -o comm 1
systemd

```
So that was easy. 
I have tried to shutdown the LS by calling `poweroff` but it was showing the same exact behavior, going to EM mode and peeping constantly. So I tried halt instead.

Now, after calling `halt`, the system shuts down thanks to your micon_shutdown.py script. No reboot into EM anymore. Only thing odd is the green power LED that is still on after everything is shut down. No fan or HDD running, just the LED still on. Wouldn't bother me too much, unfortunately, after about 3 minutes, the LED turns off and the LS starts beeping again, which can be turned off by just touching the power button.  

Checking the log flies reveals the last log entries after the "halt" command was called:
```
root@tatooine:~# less /var/log/syslog
Apr  9 13:57:25 tatooine systemd[1]: Stopping Session 1 of user root.
Apr  9 13:57:25 tatooine systemd[1]: Removed slice system-getty.slice.
Apr  9 13:57:25 tatooine systemd[1]: Removed slice system-modprobe.slice.
Apr  9 13:57:25 tatooine systemd[1]: Stopped target Graphical Interface.
Apr  9 13:57:25 tatooine systemd[1]: Stopped target Multi-User System.
Apr  9 13:57:25 tatooine systemd[1]: Stopped target Login Prompts.
Apr  9 13:57:25 tatooine systemd[1]: Stopped target Timers.
Apr  9 13:57:25 tatooine systemd[1]: apt-daily-upgrade.timer: Succeeded.
Apr  9 13:57:25 tatooine systemd[1]: Stopped Daily apt upgrade and clean activities.
Apr  9 13:57:25 tatooine systemd[1]: apt-daily.timer: Succeeded.
Apr  9 13:57:25 tatooine systemd[1]: Stopped Daily apt download activities.
Apr  9 13:57:25 tatooine systemd[1]: e2scrub_all.timer: Succeeded.
Apr  9 13:57:25 tatooine systemd[1]: Stopped Periodic ext4 Online Metadata Check for All Filesystems.
Apr  9 13:57:25 tatooine systemd[1]: logrotate.timer: Succeeded.
Apr  9 13:57:25 tatooine systemd[1]: Stopped Daily rotation of log files.
Apr  9 13:57:25 tatooine systemd[1]: exim4-base.timer: Succeeded.
Apr  9 13:57:25 tatooine systemd[1]: Stopped Daily exim4-base housekeeping.
Apr  9 13:57:25 tatooine systemd[1]: man-db.timer: Succeeded.
Apr  9 13:57:25 tatooine systemd[1]: Stopped Daily man-db regeneration.
Apr  9 13:57:25 tatooine systemd[1]: systemd-tmpfiles-clean.timer: Succeeded.
Apr  9 13:57:25 tatooine systemd[1]: Stopped Daily Cleanup of Temporary Directories.
Apr  9 13:57:25 tatooine systemd[1]: Stopping Regular background program processing daemon...
Apr  9 13:57:25 tatooine systemd[1]: Stopping LSB: exim Mail Transport Agent...
Apr  9 13:57:25 tatooine systemd[1]: Stopping LSB: Daemon for Linkstation/Kuro micro controller...

```
So if I read this correctly, other than the LED and the peeping, your script seems to be working fine for `halt`.
Do you see any issue here? Anything I may have missed?

Thanks again for the help! It's great to have people like you maintaining things like this.
I think I will try out your installer scrip as soon as I have some more time at hand. Looks promising.

Please let me know in case you want me to test anything.

Thanks and 73
Daniel

**Reply by 1000001101000** [reply]
It sounds like the script isn't running. What you describe is what I would expect from a normal halt.

I think the first step is to make sure the script is working.
try running `/lib/systemd/system-shutdown/micon_shutdown.py poweroff` and see what happens.

**Reply by DanOnDeb** [reply]
You're right, sorry. Should have checked that earlier.

```
root@tatooine:~# /lib/systemd/system-shutdown/micon_shutdown.py poweroff
Traceback (most recent call last):
  File "/lib/systemd/system-shutdown/micon_shutdown.py", line 3, in <module>
    import libmicon
ModuleNotFoundError: No module named 'libmicon'

```
I may have not installed enough of your scripts, I gather.

**Reply by 1000001101000** [reply]
lol, I forgot the most important part!

Try this:
1. Grab libmicon from https://github.com/1000001101000/Python_buffalo_libmicon and put it in /usr/local/bin
2. move micon_shutdown.py to /usr/local/bin
3. link micon_shutdown.py to the system-shutdown directory `ln -s /usr/local/bin/micon_shutdown.py /lib/systemd/system-shutdown/micon_shutdown.py`

Then try again

**Reply by DanOnDeb** [reply]
I cannot believe it. It just works. After so many hours of trying to get it fixed, it finally works.
Thank you so much! Really appreciate it. I am one happy camper now!
Please let me know in case you need my help with anything!

73
DanOnDeb

**Reply by 1000001101000** [reply]
Great!
